### PR TITLE
#48 track health on component CR

### LINKED ISSUE #48: Track health on component CR

### PR DESCRIPTION
**Pull request by jelemux**
Resolves #48

### CODE REVIEW

File: pkg/health/applicationFinder.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to list deployments for component %q: %w", componentName, err))
+	}
+
+	statefulSetList, err := af.appsClient.StatefulSets(namespace).List(ctx, metav1.ListOptions{LabelSelector: componentLabelSelector})
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to list stateful sets for component %q: %w", componentName, err))
+	}
+
+	daemonSetList, err := af.appsClient.DaemonSets(namespace).List(ctx, metav1.ListOptions{LabelSelector: componentLabelSelector})
```
Comment (cesmarvin) [inline pkg/health/applicationFinder.go:35]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>af.appsClient.DaemonSets undefined (type appsV1Client has no field or method DaemonSets)

File: pkg/health/componentRepo.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	v1 "github.com/cloudogu/k8s-component-operator/pkg/api/v1"
+	"github.com/cloudogu/k8s-component-operator/pkg/retry"
+)
+
+type defaultComponentRepo struct {
+	client componentClient
+}
+
+func (cr *defaultComponentRepo) get(ctx context.Context, name string) (*v1.Component, error) {
+	component, err := cr.client.Get(ctx, name, metav1.GetOptions{})
```
Comment (cesmarvin) [inline pkg/health/componentRepo.go:18]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>cr.client.Get undefined (type componentClient has no field or method Get)

File: pkg/health/componentRepo.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+func (cr *defaultComponentRepo) updateHealthStatus(ctx context.Context, component *v1.Component, status v1.HealthStatus) error {
+	return retry.OnConflict(func() error {
+		component, err := cr.get(ctx, component.Name)
+		if err != nil {
+			return err
+		}
+
+		component.Status.Health = status
+
+		_, err = cr.client.UpdateStatus(ctx, component, metav1.UpdateOptions{})
```
Comment (cesmarvin) [inline pkg/health/componentRepo.go:35]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>cr.client.UpdateStatus undefined (type componentClient has no field or method UpdateStatus)

File: pkg/health/deploymentController.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+type deploymentReconciler struct {
+	clientSet ecosystemClientSet
+	manager   componentManager
+}
+
+func (dr *deploymentReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := log.FromContext(ctx)
+
+	deployment, err := dr.clientSet.
+		AppsV1().Deployments(request.Namespace).
```
Comment (cesmarvin) [inline pkg/health/deploymentController.go:24]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>dr.clientSet.AppsV1 undefined (type ecosystemClientSet has no field or method AppsV1)

File: pkg/health/manager.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+)
+
+type defaultManager struct {
+	applicationFinder
+	componentRepo
+}
+
+func newManager(namespace string, clientSet ecosystemClientSet) *defaultManager {
+	return &defaultManager{
+		applicationFinder: &defaultApplicationFinder{appsClient: clientSet.AppsV1()},
```
Comment (cesmarvin) [inline pkg/health/manager.go:21]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>clientSet.AppsV1 undefined (type ecosystemClientSet has no field or method AppsV1)

File: pkg/health/manager.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		componentRepo:     &defaultComponentRepo{client: clientSet.ComponentV1Alpha1().Components(namespace)},
```
Comment (cesmarvin) [inline pkg/health/manager.go:22]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>clientSet.ComponentV1Alpha1 undefined (type ecosystemClientSet has no field or method ComponentV1Alpha1)

File: pkg/health/statefulSetController.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+type statefulSetReconciler struct {
+	clientSet ecosystemClientSet
+	manager   componentManager
+}
+
+func (ssr *statefulSetReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := log.FromContext(ctx)
+
+	statefulSet, err := ssr.clientSet.
+		AppsV1().StatefulSets(request.Namespace).
```
Comment (cesmarvin) [inline pkg/health/statefulSetController.go:24]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>ssr.clientSet.AppsV1 undefined (type ecosystemClientSet has no field or method AppsV1)

File: pkg/labels/postrenderer.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+		unstructuredConverter:  runtime.DefaultUnstructuredConverter,
+		serializer:             yamlutil.NewSerializer(),
+		labels:                 labels,
+	}
+}
+
+func (c *PostRenderer) Run(renderedManifests *bytes.Buffer) (modifiedManifests *bytes.Buffer, err error) {
+	modifiedManifests = new(bytes.Buffer)
+
+	c.documentSplitter.WithReader(renderedManifests)
```
Comment (cesmarvin) [inline pkg/labels/postrenderer.go:36]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>c.documentSplitter.WithReader undefined (type documentSplitter has no field or method WithReader)

File: pkg/labels/postrenderer.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+	for c.documentSplitter.Next() {
```
Comment (cesmarvin) [inline pkg/labels/postrenderer.go:37]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>c.documentSplitter.Next undefined (type documentSplitter has no field or method Next)

File: pkg/labels/postrenderer.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+		obj, _, err := c.unstructuredSerializer.Decode(c.documentSplitter.Bytes(), nil, nil)
```
Comment (cesmarvin) [inline pkg/labels/postrenderer.go:38]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>c.unstructuredSerializer.Decode undefined (type unstructuredSerializer has no field or method Decode)

File: pkg/labels/postrenderer.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse yaml resources: %w", err)
+		}
+
+		unstructuredMap, err := c.unstructuredConverter.ToUnstructured(obj)
```
Comment (cesmarvin) [inline pkg/labels/postrenderer.go:43]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>c.unstructuredConverter.ToUnstructured undefined (type unstructuredConverter has no field or method ToUnstructured)

File: pkg/labels/postrenderer.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+		k8sObject := &unstructured.Unstructured{Object: unstructuredMap}
+
+		originalLabels := k8sObject.GetLabels()
+		mergedLabels := make(map[string]string, len(c.labels)+len(originalLabels))
+		maps.Copy(mergedLabels, originalLabels)
+		maps.Copy(mergedLabels, c.labels)
+		k8sObject.SetLabels(mergedLabels)
+
+		yamlBytes, err := c.serializer.Marshal(k8sObject)
```
Comment (cesmarvin) [inline pkg/labels/postrenderer.go:56]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>c.serializer.Marshal undefined (type genericYamlSerializer has no field or method Marshal)

File: pkg/labels/postrenderer.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to serialize resources back to yaml: %w", err)
+		}
+
+		modifiedManifests.Write(yamlBytes)
+		modifiedManifests.WriteString("\n---\n")
+	}
+	if err = c.documentSplitter.Err(); err != nil {
```
Comment (cesmarvin) [inline pkg/labels/postrenderer.go:64]: 🚫 **[checkstyle]** <typecheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>c.documentSplitter.Err undefined (type documentSplitter has no field or method Err)

File: pkg/health/applicationFinder.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+	appsClient appsV1Client
+}
+
+func (af *defaultApplicationFinder) findComponentApplications(ctx context.Context, componentName string, namespace string) (*appsv1.DeploymentList, *appsv1.StatefulSetList, *appsv1.DaemonSetList, error) {
+	componentLabelSelector := labels.ValidatedSetSelector{
+		v1.ComponentNameLabelKey: componentName,
+	}.String()
+
+	var errs []error
+	deploymentList, err := af.appsClient.Deployments(namespace).List(ctx, metav1.ListOptions{LabelSelector: componentLabelSelector})
```
[the remark this answers is not in the document]
Comment (jelemux) [reply pkg/health/applicationFinder.go:25]: WTF, these are all methods that are very much defined

File: pkg/health/daemonSetController.go [hunk 15]
```diff
@@ -0,0 +1,53 @@
+package health
```
[the remark this answers is not in the document]
Comment (jelemux) [reply pkg/health/daemonSetController.go:1]: Gonna ignore this as we'd have to use Generics to resolve it.

File: pkg/health/deploymentController.go [hunk 16]
```diff
@@ -0,0 +1,53 @@
+package health
```
[the remark this answers is not in the document]
Comment (jelemux) [reply pkg/health/deploymentController.go:1]: Gonna ignore this as we'd have to use Generics to resolve it.

File: pkg/health/statefulSetController.go [hunk 17]
```diff
@@ -0,0 +1,53 @@
+package health
```
[the remark this answers is not in the document]
Comment (jelemux) [reply pkg/health/statefulSetController.go:1]: Gonna ignore this as we'd have to use Generics to resolve it.

**Review by alexander-dammeier — CHANGES_REQUESTED** [review]
Just a little question to the testdata but everything else LGTM :)

File: pkg/labels/testdata/longhorn.yaml [hunk 18]
```diff
@@ -0,0 +1,4538 @@
...
+    
+    
+    
+    
+    
+    
+    
+    
+    
+
```
[the remark this answers is not in the document]
Comment (jelemux) [reply pkg/labels/testdata/longhorn.yaml:82]: It's what the output of `helm template` for longhorn looks like. So yeah, it's definitely not bad data for testing :)

File: pkg/controllers/interfaces.go [hunk 19]
```diff
@@ -73,3 +79,15 @@ type requeuableError interface {
...
+//goland:noinspection GoUnusedType
+type appsV1Interface interface {
+	appsv1.AppsV1Interface
+}
+
+//nolint:unused
+//goland:noinspection GoUnusedType
+type componentV1Alpha1Interface interface {
+	ecosystem.ComponentV1Alpha1Interface
+}
```
Comment (alexander-dammeier) [inline pkg/controllers/interfaces.go:93]: good to know that our components are in an alpha phase^^

File: pkg/health/changeEventFilter.go [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+func (h *healthChangeEventFilter) Update(event event.UpdateEvent) bool {
+	if !hasComponentLabel(event.ObjectNew) {
+		return false
+	}
+
+	switch oldObj := event.ObjectOld.(type) {
+	case *appsv1.Deployment:
+		newObj := event.ObjectNew.(*appsv1.Deployment)
```
Comment (cesmarvin) [inline pkg/health/changeEventFilter.go:25]: 🚫 **[checkstyle]** <forcetypeassert> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>type assertion must be checked

File: pkg/health/changeEventFilter.go [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+		return oldObj.Spec.Replicas != newObj.Spec.Replicas ||
+			oldObj.Status.Replicas != newObj.Status.Replicas ||
+			oldObj.Status.UpdatedReplicas != newObj.Status.UpdatedReplicas ||
+			oldObj.Status.AvailableReplicas != newObj.Status.AvailableReplicas
+	case *appsv1.StatefulSet:
+		newObj := event.ObjectNew.(*appsv1.StatefulSet)
```
Comment (cesmarvin) [inline pkg/health/changeEventFilter.go:31]: 🚫 **[checkstyle]** <forcetypeassert> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>type assertion must be checked

File: pkg/health/changeEventFilter.go [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+		return oldObj.Spec.Replicas != newObj.Spec.Replicas ||
+			oldObj.Status.Replicas != newObj.Status.Replicas ||
+			oldObj.Status.UpdatedReplicas != newObj.Status.UpdatedReplicas ||
+			oldObj.Status.AvailableReplicas != newObj.Status.AvailableReplicas
+	case *appsv1.DaemonSet:
+		newObj := event.ObjectNew.(*appsv1.DaemonSet)
```
Comment (cesmarvin) [inline pkg/health/changeEventFilter.go:37]: 🚫 **[checkstyle]** <forcetypeassert> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>type assertion must be checked

File: pkg/health/daemonSetController.go [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
```
Comment (cesmarvin) [inline pkg/health/daemonSetController.go:1]: 🚫 **[checkstyle]** <dupl> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>1-51 lines are duplicate of `pkg/health/deploymentController.go:1-51`

File: pkg/health/deploymentController.go [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
```
Comment (cesmarvin) [inline pkg/health/deploymentController.go:1]: 🚫 **[checkstyle]** <dupl> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>1-51 lines are duplicate of `pkg/health/statefulSetController.go:1-51`

File: pkg/health/statefulSetController.go [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
```
Comment (cesmarvin) [inline pkg/health/statefulSetController.go:1]: 🚫 **[checkstyle]** <dupl> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>1-51 lines are duplicate of `pkg/health/daemonSetController.go:1-51`

**Comment by alexander-dammeier** [comment]
lgtm

**Pull request merged**